frontend/src/app/main/ui/measurements.cljs

```diff
@@ -475,25 +468,469 @@
+                     val            (int (max (+ @original-value (/ delta zoom)) 0))
+                     layout-item-margin (cond
+                                          hover-all? (assoc margin :m1 val :m2 val :m3 val :m4 val)
+                                          hover-v?   (assoc margin :m1 val :m3 val)
+                                          hover-h?   (assoc margin :m2 val :m4 val)
+                                          :else      (assoc margin margin-num val))
+                     layout-item-margin-type (if (= (:m1 margin) (:m2 margin) (:m3 margin) (:m4 margin)) :simple :multiple)
+                     modifiers      (dwm/create-modif-tree [shape-id]
+                                                           (-> (ctm/empty)
+                                                               (ctm/change-property  :layout-item-margin layout-item-margin)
+                                                               (ctm/change-property  :layout-item-margin-type layout-item-margin-type)))]
+                 (st/emit! (dwm/set-modifiers modifiers)))))))]
+
+
+    [:*
+     [:rect.margin-rect {:x (:x rect-data)
+                          :y (:y rect-data)
+                          :width (:width rect-data)
+                          :height (:height rect-data)
+                          :on-mouse-enter on-mouse-enter
+                          :on-mouse-leave on-mouse-leave
+                          :on-pointer-down on-pointer-down
+                          :on-lost-pointer-capture on-lost-pointer-capture
+                          :on-mouse-move on-mouse-move
+                          :style {:fill (if (or hover? selected?) warning-color "none")
+                                  :cursor (when (or hover? selected?)
+                                            (if (= (:resize-axis rect-data) :x) (cur/resize-ew 0) (cur/resize-ew 90)))
+                                  :opacity (if selected? 0.5 0.25)}}]]))
+
+(mf/defc margin-rects [{:keys [shape frame zoom alt? shift?]}]
+  (let [shape-id                   (:id shape)
+        margins-selected           (mf/deref refs/workspace-margins-selected)
+        hover-value                (mf/use-var 0)
+        mouse-pos                  (mf/use-var nil)
+        hover                      (mf/use-var nil)
+        hover-all?                 (and (not (nil? @hover)) alt?)
+        hover-v?                   (and (or (= @hover :m1) (= @hover :m3)) shift?)
+        hover-h?                   (and (or (= @hover :m2) (= @hover :m4)) shift?)
+        margin                    (:layout-item-margin shape)
+        [width height x1 x2 y1 y2] ((juxt :width :height :x1 :x2 :y1 :y2) (:selrect shape))
+        on-mouse-enter            (fn [hover-type val]
+                                    (reset! hover hover-type)
+                                    (reset! hover-value val))
+        on-mouse-leave             #(reset! hover nil)
+
+        margin-display-data       [{:shape-id shape-id
```

Comment:
I think there are better ways to do this... there is a lot of repetition. Also, passing this object directly to the component is not good idea.